group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,390 @@
+                        .add(validCurrentAssignment.get(retainedPartitionsCount).partition());
+                    allAssignedStickyPartitions.add(validCurrentAssignment.get(retainedPartitionsCount));
+                    remainingMembersToGetAnExtraPartition--;
+                }
+            }
+
+            if (remaining >= 0) {
+                potentiallyUnfilledMembers.put(memberId, remaining);
+            }
+
+        });
+
+        return allAssignedStickyPartitions;
+    }
+
+    /**
+     * Filters the current assignment of partitions for a given member.
+     *
+     * If a partition is assigned to a member not subscribed to its topic or
+     * if the rack-aware strategy is to be used but there is a mismatch,
+     * the partition is excluded from the valid assignment and stored for future consideration.
+     *
+     * @param memberId              The Id of the member whose assignment is being validated.
+     * @param assignedPartitions    The partitions currently assigned to the member.
+     *
+     * @return List of valid partitions after applying the filters.
+     */
+    private List<TopicIdPartition> validCurrentAssignment(
+        String memberId,
+        Map<Uuid, Set<Integer>> assignedPartitions
+    ) {
+        List<TopicIdPartition> validCurrentAssignmentList = new ArrayList<>();
+        Set<Uuid> subscriptionSet = new HashSet<>(subscriptionList);
+
+        assignedPartitions.forEach((topicId, currentAssignment) -> {
+            if (subscriptionSet.contains(topicId)) {
+                currentAssignment.forEach((partition) -> {
+                    TopicIdPartition topicIdPartition = new TopicIdPartition(topicId, partition);
+                    if (useRackAwareStrategy && rackInfo.racksMismatch(memberId, topicIdPartition)) {
+                        currentPartitionOwners.put(topicIdPartition, memberId);
+                    } else {
+                        validCurrentAssignmentList.add(topicIdPartition);
+                    }
+                });
+            }
+        });
+
+        return validCurrentAssignmentList;
+    }
+
+    /**
+     * This method iterates over the unassigned partitions and attempts to allocate them
+     * to members while considering their rack affiliations.
+     */
+    private void rackAwareRoundRobinAssignment() {
+        Queue<String> roundRobinMembers = new LinkedList<>(unfilledMembers.keySet());
+
+        // Sorts partitions in ascending order by number of potential consumers with matching racks.
+        // Partitions with no potential members aren't included in this list.
+        List<TopicIdPartition> sortedPartitions = rackInfo.sortPartitionsByRackConsumers(unassignedPartitions);
+
+        sortedPartitions.forEach((partition) -> {
+            boolean assigned = false;
+            for (int i = 0; i < roundRobinMembers.size() && !assigned; i++) {
```

Comment:
`while(!roundRobinMembers.isEmpty() && !assigned)` is more readable since most for loops do not change the size

Reply:
I don't want the loop to run until the roundRobinMembers list is empty though, this could cause an infinite loop. I just want it to go through every member at the very least and then if it doesn't find a member that is suitable then we just move on and the partition might get assigned later

Reply:
the while loop will still iterate through all members at the very least right? since we poll one every iteration

can you help me understand why the while loop can be infinite whereas the for loop cannot?

Reply:
because we keep adding members back to the round robin queue until the remaining partitions to be received by the member hits 0. Therefore, the roundRobinMembers queue will never be empty unless all the partitions are assigned/ all the quotas are met.

Reply:
Here we might assign all the partitions that have matching racks but there's gonna be other partitions that haven't been assigned yet and therefore the remaining number of partitions to be received by members may not have reached zero yet.

Reply:
the `&& !assigned` part would prevent the infinite loop, no?

Reply:
discussed offline. makes sense